plugins/storage/fileshare/storagefsvm/src/main/java/org/apache/cloudstack/storage/fileshare/lifecycle/StorageFsVmFileShareLifeCycle.java

```diff
@@ -0,0 +1,329 @@
+        final String filesystem = "{{ fsvm.filesystem }}";
+        fsVmConfig = fsVmConfig.replace(filesystem, fileSystem);
+        return fsVmConfig;
+    }
+
+    private String getStorageFsVmName(String fileShareName) {
+        String prefix = String.format("%s-%s", FileShareVmNamePrefix, fileShareName);
+        String suffix = Long.toHexString(System.currentTimeMillis());
+
+        if (!NetUtils.verifyDomainNameLabel(prefix, true)) {
+            prefix = prefix.replaceAll("[^a-zA-Z0-9-]", "");
+        }
+        int nameLength = prefix.length() + suffix.length() + FileShareVmNamePrefix.length();
+        if (nameLength > 63) {
+            int prefixLength = prefix.length() - (nameLength - 63);
+            prefix = prefix.substring(0, prefixLength);
+        }
+        return (String.format("%s-%s", prefix, suffix));
+    }
+
+    private UserVm createFileShareVM(Long zoneId, Account owner, List<Long> networkIds, String name, Long serviceOfferingId, Long diskOfferingId, FileShare.FileSystemType fileSystem, Long size, Long minIops, Long maxIops) throws ResourceUnavailableException, InsufficientCapacityException, ResourceAllocationException {
+        ServiceOfferingVO serviceOffering = serviceOfferingDao.findById(serviceOfferingId);
+
+        Long diskSize = null;
+        if (diskOfferingId != null) {
+            DiskOfferingVO diskOffering = diskOfferingDao.findById(diskOfferingId);
+            if (diskOffering.isCustomized()) {
+                diskSize = size;
+            }
+        }
+
+        DataCenter zone = entityMgr.findById(DataCenter.class, zoneId);
+        Hypervisor.HypervisorType availableHypervisor = resourceMgr.getAvailableHypervisor(zoneId);
+        VMTemplateVO template = templateDao.findSystemVMReadyTemplate(zoneId, availableHypervisor);
+
+        LaunchPermissionVO existingPermission = launchPermissionDao.findByTemplateAndAccount(template.getId(), owner.getId());
+        if (existingPermission == null) {
+            LaunchPermissionVO launchPermission = new LaunchPermissionVO(template.getId(), owner.getId());
+            launchPermissionDao.persist(launchPermission);
+        }
+
+        String hostName = getStorageFsVmName(name);
+
+        Network.IpAddresses addrs = new Network.IpAddresses(null, null);
+        Map<String, String> customParameterMap = new HashMap<String, String>();
+        if (minIops != null) {
+            customParameterMap.put("minIopsDo", minIops.toString());
+            customParameterMap.put("maxIopsDo", maxIops.toString());
+        }
+        List<String> keypairs = new ArrayList<String>();
+
+        UserVm vm;
+        String fsVmConfig = getStorageFsVmConfig(fileSystem.toString().toLowerCase());
+        String base64UserData = Base64.encodeBase64String(fsVmConfig.getBytes(com.cloud.utils.StringUtils.getPreferredCharset()));
+        vm = userVmService.createAdvancedVirtualMachine(zone, serviceOffering, template, networkIds, owner, hostName, hostName,
```

Comment:
maybe check the zone type here ?

(only advanced zone without SG is supported, right ?)

Reply:
Good idea. Added to validateAndGetZone in FileShareServiceImpl.java